### PR TITLE
Feature/mozilla tts

### PR DESCRIPTION
**Pull request by domcross**
## Description
Implement TTS module for Mozilla-TTS server

## How to test
Fire up Mozilla-TTS server, set mycroft.conf: tts.module "mozilla" and configure tts.mozilla.url, mycroft-speak "My name is Mycroft and I am funky"

## Contributor license agreement signed?
CLA [X] (Whether you have signed a [CLA - Contributor Licensing Agreement](https://mycroft.ai/cla/)

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by el-tocino** [comment]
The dev server moz tts includes can be accessed similarly to mimic2, but no visemes are returned.  A slightly more generic tool that connects to a url and expects a wav file in return might be ideal.

File: mycroft/tts/mozilla_tts.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        if config is None:
+            config = Configuration.get().get("tts", {}).get("mozilla", {})
+        super(MozillaTTS, self).__init__(lang, config,
+            MozillaTTSValidator(self))
+        self.url = config['url']
+        self.type = 'wav'
+
+    def get_tts(self, sentence, wav_file):
+        wav_name = hashlib.sha1(sentence.encode('utf-8')).hexdigest() + ".wav"
+        wav_file = "/tmp/" + wav_name
```
Comment (el-tocino) [inline mycroft/tts/mozilla_tts.py:49]: there's typically a /tmp/mycroft/ folder that might be better used for this.

Comment (forslund) [reply mycroft/tts/mozilla_tts.py:49]: I think you should be able to use wav_file without modification that should contain the complete path el-tocino mentions. If this is changed the use of local caching won't work.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by forslund** [comment]
Very cool! Are there any good instructions for setting up mozilla TTS server outthere?

**Comment by domcross** [comment]
> Very cool! Are there any good instructions for setting up mozilla TTS server outthere?

When you have a pre-trained model you can run `server.py` from [TTS\server](https://github.com/mozilla/TTS/tree/dev/TTS/server) which is based on Flask. There are also instructions for Nginx and build a server package with the model embedded.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by JarbasAl** [comment]
there is also https://github.com/synesthesiam/docker-mozillatts

using the pip .whl for the pre trained model [here](https://github.com/mozilla/TTS/wiki/Released-Models#simple-packaging---self-contained-package-that-runs-an-http-api-for-a-pre-trained-tts-model) did not work for me, i needed to update numpy and downgrade numba for it to work

**Review by krisgesling — COMMENTED** [review]
Nice one! You beat me to it :)

We are going to be extracting all the TTS and STT modules out to plugins, but given the work happening on voices right now I think it makes sense to merge this in.

Outside the scope of this PR, but inspired by it - I'm wondering if we should extract out the phrase chunking functions that are currently in the Mimic2 module so that any TTS module can make use of them?

**Comment by domcross** [comment]
> I'm wondering if we should extract out the phrase chunking functions that are currently in the Mimic2 module so that any TTS module can make use of them?

I have seen these functions and left them out intentionally as Mozilla-TTS server implements phrase chunking functions itself.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by forslund** [comment]
Finally got the TTS setup, it's working very well!

I could minimize the get_tts method to

```python
   def get_tts(self, sentence, wav_file):                                      
        req_route = self.url + sentence                                         
        response = requests.get(req_route)                                      
        with open(wav_file, 'wb') as f:                                         
            f.write(response.content)                                           
        return (wav_file, None)  # No phonemes
```

This lets mycroft take care of all caching instead of duplicating the check here.

Another way to make the code cleaner could be to use the params argument in requests.get

setting the url in config to `http://0.0.0.0:5002/api/tts`

then doing the request call using `requests.get(self.url, params={'text': sentence})`

This would automatically escape anything in `sentence?` that could mess up the url.

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

File: mycroft/tts/mozilla_tts.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# limitations under the License.
+#
+
+import hashlib
+import os
+import requests
+
+from .tts import TTS, TTSValidator
+from mycroft.configuration import Configuration
+from mycroft.util.log import LOG
```
Comment (forslund) [inline mycroft/tts/mozilla_tts.py:22]: Seems like now os, hashlib and mycroft.util.log isn't used. Apart from those tiny things it looks great!

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2713))

**Comment by domcross** [comment]
in case this PR is merged adding the "hacktoberfest-accepted" label to it would be a cool thing ;-)

**Review by forslund — APPROVED** [review]
Works excellent now! Can without issue say "Sharky & George"

**Review by JarbasAl — COMMENTED** [review]
just noticed i made an "invisible" review, @chance pointed out comments dont appear until i submit the review....

doesnt matter now that it is merged, i opened a new PR with the changes i requested. most changes were addressed either way.

I wonder how many times i thought i reviewed something, but not really....

**Comment by krisgesling** [comment]
Instructions have been added to the docs:
https://mycroft-ai.gitbook.io/docs/using-mycroft-ai/customizations/tts-engine#mozilla-tts

**Pull request merged**